packages/contracts/test/attestation.test.ts

```diff
@@ -156,4 +158,791 @@ describe('Attestations', function () {
             )
     })
+
+    it('should fail to build hashchain without attestation', async () => {
+        const accounts = await ethers.getSigners()
+        const attester = accounts[2]
+        await unirepContract
+            .connect(attester)
+            .attesterSignUp(EPOCH_LENGTH)
+            .then((t) => t.wait())
+
+        const epoch = await unirepContract.attesterCurrentEpoch(
+            attester.address
+        )
+        await expect(unirepContract.buildHashchain(attester.address, epoch)).to
+            .be.reverted
+    })
+
+    it('should build correct hash chain with different epoch keys', async () => {
+        const accounts = await ethers.getSigners()
+        const attester = accounts[2]
+
+        const epoch = await unirepContract.attesterCurrentEpoch(
+            attester.address
+        )
+        for (let i = 0; i < 5; i++) {
+            const epochKey = genRandomSalt() % BigInt(2 ** EPOCH_TREE_DEPTH)
+            const posRep = Math.floor(Math.random() * 10)
+            const negRep = Math.floor(Math.random() * 10)
+            const graffiti = genRandomSalt()
+            let timestamp
+            {
+                const tx = await unirepContract
+                    .connect(attester)
+                    .submitAttestation(
+                        epoch,
+                        epochKey,
+                        posRep,
+                        negRep,
+                        graffiti
+                    )
+                const receipt = await tx.wait()
+                const block = await ethers.provider.getBlock(
+                    receipt.blockNumber
+                )
+                timestamp = block.timestamp
+            }
+
+            {
+                const tx = await unirepContract.buildHashchain(
+                    attester.address,
+                    epoch
+                )
+                await tx.wait()
+            }
+
+            const currentHead = hash3([attester.address, epoch, i])
+            const head = hash6([
+                currentHead,
+                epochKey,
+                posRep,
+                negRep,
+                graffiti,
+                timestamp,
+            ])
+            const _attesterId = await unirepContract.hashchainMapping(head, 0)
+            const _epoch = await unirepContract.hashchainMapping(head, 1)
+            const _index = await unirepContract.hashchainMapping(head, 2)
+            expect(_attesterId.toString()).to.equal(
+                BigInt(attester.address).toString()
+            )
+            expect(_epoch.toString()).to.equal(epoch.toString())
+            expect(_index.toNumber()).to.equal(i)
+            const count = await unirepContract.attesterHashchainTotalCount(
+                attester.address,
+                epoch
+            )
+            expect(count).to.equal(i + 1)
+        }
+    })
+
+    it('should build correct hash chain with the same epoch key', async () => {
```

Comment:
Also I think we should evaluate this based on codepaths. If we're testing a codebpath here that we're not hitting elsewhere then I'm all for it.

Reply:
Actually it is for this: https://github.com/Unirep/Unirep/pull/186
but it was fixed
I am not sure if we should keep this

> It would be good to build 5 hashchains each containing 1 attestation, and then verify that the final balance output is correct (this may be in another test).

Yeah we can build another tests for this to check the final balance

Reply:
Right maybe it's good to keep this as separate test.